make all failed

**Describe the bug**
run `make all`

```shell
go mod vendor; \
    /home/ubuntu/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" rbac:roleName=manager-role crd webhook paths="./vendor/github.com/gocrane/api/..." output:crd:artifacts:config=deploy/manifests; \
    rm -rf vendor
bash: line 1: /home/ubuntu/go/bin/controller-gen: No such file or directory
make: *** [Makefile:64: manifests] Error 127
```

it can not find the command `controller-gen`, so I check the `Makefile`.

```makefile
controller-gen:
ifeq (, $(shell which controller-gen))
        @{ \
        set -e ;\
        CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
        cd $$CONTROLLER_GEN_TMP_DIR ;\
        go mod init tmp ;\
        go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0 ;\
        rm -rf $$CONTROLLER_GEN_TMP_DIR ;\
        }
CONTROLLER_GEN=$(shell go env GOPATH)/bin/controller-gen
else
CONTROLLER_GEN=$(shell which controller-gen)
endif
```

`go get` should be modified to `go install`.

```shell
In earlier versions of Go, 'go get' was used to build and install packages.
Now, 'go get' is dedicated to adjusting dependencies in go.mod. 'go install'
may be used to build and install commands instead.
```

Also, golangci-lint is the same situation.

I modified it, and run `make all` again.

```shell

go mod vendor; \
    /home/ubuntu/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" rbac:roleName=manager-role crd webhook paths="./vendor/github.com/gocrane/api/..." output:crd:artifacts:config=deploy/manifests; \
    rm -rf vendor
go generate ./...
go fmt ./...
pkg/metricquery/type.go:98:29: invalid operation: cannot compare m.Workload.Selector == nil (operator == not defined on untyped nil) (typecheck)
		if m.Workload.Selector == nil {
		                          ^
pkg/metricquery/type.go:126:3: undeclared name: `klog` (typecheck)
		klog.Errorf("Failed to build unique key, validate metric err: %v", err)
		^
pkg/metricquery/type.go:141:3: undeclared name: `klog` (typecheck)
		klog.Errorf("Failed to build unique key, not supported metric type %v", m.Type)
		^
pkg/metricquery/type.go:148:28: invalid operation: cannot compare m.Workload.Selector != nil (operator != not defined on untyped nil) (typecheck)
	if m.Workload.Selector != nil {
	                          ^
pkg/metricquery/type.go:163:29: invalid operation: cannot compare m.Container.Selector != nil (operator != not defined on untyped nil) (typecheck)
	if m.Container.Selector != nil {
	                           ^
pkg/metricquery/type.go:177:23: invalid operation: cannot compare m.Pod.Selector != nil (operator != not defined on untyped nil) (typecheck)
	if m.Pod.Selector != nil {
	                     ^
pkg/metricquery/type.go:189:24: invalid operation: cannot compare m.Node.Selector != nil (operator != not defined on untyped nil) (typecheck)
	if m.Node.Selector != nil {
	                      ^
pkg/metricquery/type.go:201:24: invalid operation: cannot compare m.Prom.Selector != nil (operator != not defined on untyped nil) (typecheck)
	if m.Prom.Selector != nil {
	                      ^
pkg/metricquery/type.go:8:2: "k8s.io/klog/v2" imported but not used (typecheck)
	"k8s.io/klog/v2"
	^
pkg/providers/proxy.go:42:4: r.Lock undefined (type *RealTimeDataProxy has no field or method Lock) (typecheck)
	r.Lock()
	  ^
pkg/providers/proxy.go:43:10: r.Unlock undefined (type *RealTimeDataProxy has no field or method Unlock) (typecheck)
	defer r.Unlock()
	        ^
pkg/providers/proxy.go:48:4: r.Lock undefined (type *RealTimeDataProxy has no field or method Lock) (typecheck)
	r.Lock()
	  ^
pkg/providers/proxy.go:49:10: r.Unlock undefined (type *RealTimeDataProxy has no field or method Unlock) (typecheck)
	defer r.Unlock()
	        ^
pkg/providers/proxy.go:54:4: r.Lock undefined (type *RealTimeDataProxy has no field or method Lock) (typecheck)
	r.Lock()
	  ^
pkg/providers/proxy.go:55:10: r.Unlock undefined (type *RealTimeDataProxy has no field or method Unlock) (typecheck)
	defer r.Unlock()
	        ^
pkg/providers/proxy.go:97:4: h.Lock undefined (type *HistoryDataProxy has no field or method Lock) (typecheck)
	h.Lock()
	  ^
pkg/providers/proxy.go:98:10: h.Unlock undefined (type *HistoryDataProxy has no field or method Unlock) (typecheck)
	defer h.Unlock()
	        ^
pkg/providers/proxy.go:106:4: h.Lock undefined (type *HistoryDataProxy has no field or method Lock) (typecheck)
	h.Lock()
	  ^
pkg/providers/proxy.go:107:10: h.Unlock undefined (type *HistoryDataProxy has no field or method Unlock) (typecheck)
	defer h.Unlock()
	        ^
pkg/providers/proxy.go:112:4: h.Lock undefined (type *HistoryDataProxy has no field or method Lock) (typecheck)
	h.Lock()
	  ^
pkg/providers/proxy.go:113:10: h.Unlock undefined (type *HistoryDataProxy has no field or method Unlock) (typecheck)
	defer h.Unlock()
	        ^
pkg/prediction/generic.go:80:2: undeclared name: `klog` (typecheck)
	klog.V(4).InfoS("Put tuple{query,caller,config} into with channel.", "query", q.MetricNamer.BuildUniqueKey(), "caller", q.Caller)
	^
pkg/prediction/generic.go:9:2: "k8s.io/klog/v2" imported but not used (typecheck)
	"k8s.io/klog/v2"
	^
pkg/ensurance/executor/podinfo/pod_info.go:8:2: could not import k8s.io/apimachinery/pkg/types (-: could not load export data: cannot import "k8s.io/apimachinery/pkg/types" (unstable iexport format version 2, just rebuild compiler and std library), export data is newer version - update tool) (typecheck)
	"k8s.io/apimachinery/pkg/types"
	^
pkg/ensurance/grpc/connection.go:8:2: "k8s.io/klog/v2" imported but not used (typecheck)
	"k8s.io/klog/v2"
	^
pkg/ensurance/analyzer/podqos-fetcher.go:12:2: could not import k8s.io/apimachinery/pkg/selection (-: could not load export data: cannot import "k8s.io/apimachinery/pkg/selection" (unstable iexport format version 2, just rebuild compiler and std library), export data is newer version - update tool) (typecheck)
	"k8s.io/apimachinery/pkg/selection"
	^
pkg/ensurance/collector/cadvisor/cadvisor_linux.go:59:17: cannot use new(CadvisorManager) (value of type *CadvisorManager) as Manager value in variable declaration: *CadvisorManager does not implement Manager (missing method GetContainerInfo) (typecheck)
var _ Manager = new(CadvisorManager)
                ^
pkg/ensurance/collector/cadvisor/cadvisor_linux.go:92:9: cannot use &(CadvisorManager literal) (value of type *CadvisorManager) as Manager value in return statement: *CadvisorManager does not implement Manager (missing method GetContainerInfo) (typecheck)
	return &CadvisorManager{
	       ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:247:5: cm.Lock undefined (type *cpuManager has no field or method Lock) (typecheck)
	cm.Lock()
	   ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:248:11: cm.Unlock undefined (type *cpuManager has no field or method Unlock) (typecheck)
	defer cm.Unlock()
	         ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:284:5: cm.Lock undefined (type *cpuManager has no field or method Lock) (typecheck)
	cm.Lock()
	   ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:285:11: cm.Unlock undefined (type *cpuManager has no field or method Unlock) (typecheck)
	defer cm.Unlock()
	         ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:462:5: cm.Lock undefined (type *cpuManager has no field or method Lock) (typecheck)
	cm.Lock()
	   ^
pkg/ensurance/cm/cpumanager/cpu_manager.go:463:11: cm.Unlock undefined (type *cpuManager has no field or method Unlock) (typecheck)
	defer cm.Unlock()
	         ^
pkg/ensurance/collector/noderesourcetopology/noderesourcetopology.go:15:13: could not import k8s.io/apimachinery/pkg/types (-: could not load export data: cannot import "k8s.io/apimachinery/pkg/types" (unstable iexport format version 2, just rebuild compiler and std library), export data is newer version - update tool) (typecheck)
	patchtypes "k8s.io/apimachinery/pkg/types"
	           ^
pkg/controller/ehpa/effective_hpa_controller.go:157:12: c.Status undefined (type *EffectiveHPAController has no field or method Status) (typecheck)
		err := c.Status().Update(ctx, ehpa)
		         ^
pkg/controller/ehpa/hpa.go:164:12: c.Update undefined (type *EffectiveHPAController has no field or method Update) (typecheck)
		err := c.Update(ctx, hpaExist)
		         ^
pkg/controller/ehpa/predict.go:90:12: c.Update undefined (type *EffectiveHPAController has no field or method Update) (typecheck)
		err := c.Update(ctx, predictionExist)
		         ^
pkg/controller/ehpa/substitute.go:93:12: c.Update undefined (type *EffectiveHPAController has no field or method Update) (typecheck)
		err := c.Update(ctx, substituteExist)
		         ^
pkg/controller/ehpa/substitute_controller.go:64:12: c.Update undefined (type *SubstituteController has no field or method Update) (typecheck)
		err := c.Update(ctx, substitute)
		         ^
pkg/controller/ehpa/substitute_controller.go:76:12: c.Status undefined (type *SubstituteController has no field or method Status) (typecheck)
		err := c.Status().Update(ctx, substitute)
		         ^
pkg/controller/ehpa/hpa_event_handler.go:4:2: "k8s.io/api/autoscaling/v2beta2" imported but not used as autoscalingv2 (typecheck)
	autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
	^
pkg/prediction/percentile/prediction.go:216:14: p.WithCh undefined (type *percentilePrediction has no field or method WithCh) (typecheck)
			qc := <-p.WithCh
			          ^
pkg/prediction/percentile/prediction.go:286:14: p.DelCh undefined (type *percentilePrediction has no field or method DelCh) (typecheck)
			qc := <-p.DelCh
			          ^
pkg/prediction/percentile/prediction.go:311:7: p.GetHistoryProvider undefined (type *percentilePrediction has no field or method GetHistoryProvider) (typecheck)
	if p.GetHistoryProvider() == nil {
	     ^
pkg/prediction/percentile/prediction.go:316:30: p.GetHistoryProvider undefined (type *percentilePrediction has no field or method GetHistoryProvider) (typecheck)
	historyTimeSeries, err := p.GetHistoryProvider().QueryTimeSeries(namer, start, end, c.sampleInterval)
	                            ^
pkg/prediction/percentile/prediction.go:391:33: p.GetRealtimeProvider undefined (type *percentilePrediction has no field or method GetRealtimeProvider) (typecheck)
	latestTimeSeriesList, err := p.GetRealtimeProvider().QueryLatestTimeSeries(namer)
	                               ^
cmd/craned/app/options/server_options.go:46:13: undeclared name: `gcfg` (typecheck)
		if err := gcfg.FatalOnly(gcfg.ReadInto(&gconfig, grafanaConfig)); err != nil {
		          ^
cmd/craned/app/options/server_options.go:72:8: o.APIKey undefined (type *ServerOptions has no field or method APIKey) (typecheck)
		if o.APIKey == "" && o.Username == "" && o.Password == "" {
		     ^
cmd/craned/app/options/server_options.go:10:2: "gopkg.in/gcfg.v1" imported but not used (typecheck)
	"gopkg.in/gcfg.v1"
	^
make: *** [Makefile:90: lint] Error 1
```

**Environment (please complete the following information):**
 - K8S Version: no
 - Crane Version: v0.8.0
 - Go Version: 1.18.5

--- a/Makefile
+++ b/Makefile
@@ -164,30 +164,9 @@ ifneq ($(REGISTRY_USER_NAME), "")
 endif
 	docker push ${ADAPTER_IMG}
 
-# go-get-tool will 'go get' any package $2 and install it to $1.
-PROJECT_DIR := $(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
-define go-get-tool
-@[ -f $(1) ] || { \
-set -e ;\
-TMP_DIR=$$(mktemp -d) ;\
-cd $$TMP_DIR ;\
-go mod init tmp ;\
-echo "Downloading $(2)" ;\
-GOBIN=$(PROJECT_DIR)/bin go get $(2) ;\
-rm -rf $$TMP_DIR ;\
-}
-endef
-
 controller-gen:
 ifeq (, $(shell which controller-gen))
-	@{ \
-	set -e ;\
-	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
-	cd $$CONTROLLER_GEN_TMP_DIR ;\
-	go mod init tmp ;\
-	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0 ;\
-	rm -rf $$CONTROLLER_GEN_TMP_DIR ;\
-	}
+	hack/install-tools.sh sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0
 CONTROLLER_GEN=$(shell go env GOPATH)/bin/controller-gen
 else
 CONTROLLER_GEN=$(shell which controller-gen)
@@ -195,15 +174,7 @@ endif
 
 golangci-lint:
 ifeq (, $(shell which golangci-lint))
-	@{ \
-	set -e ;\
-	export GO111MODULE=on; \
-	GOLANG_LINT_TMP_DIR=$$(mktemp -d) ;\
-	cd $$GOLANG_LINT_TMP_DIR ;\
-	go mod init tmp ;\
-	go get github.com/golangci/golangci-lint/cmd/golangci-lint@v1.43.0 ;\
-	rm -rf $$GOLANG_LINT_TMP_DIR ;\
-	}
+	hack/install-tools.sh github.com/golangci/golangci-lint/cmd/golangci-lint@v1.43.0
 GOLANG_LINT=$(shell go env GOPATH)/bin/golangci-lint
 else
 GOLANG_LINT=$(shell which golangci-lint)
@@ -211,15 +182,7 @@ endif
 
 goimports:
 ifeq (, $(shell which goimports))
-	@{ \
-	set -e ;\
-	export GO111MODULE=on; \
-	GO_IMPORTS_TMP_DIR=$$(mktemp -d) ;\
-	cd $$GO_IMPORTS_TMP_DIR ;\
-	go mod init tmp ;\
-	go get golang.org/x/tools/cmd/goimports@v0.1.7 ;\
-	rm -rf $$GO_IMPORTS_TMP_DIR ;\
-	}
+	hack/install-tools.sh golang.org/x/tools/cmd/goimports@v0.1.7
 GO_IMPORTS=$(shell go env GOPATH)/bin/goimports
 else
 GO_IMPORTS=$(shell which goimports)
@@ -227,16 +190,7 @@ endif
 
 mockgen:
 ifeq (, $(shell which mockgen))
-	@{ \
-	set -e ;\
-	export GO111MODULE=on; \
-	GO_MOCKGEN_TMP_DIR=$$(mktemp -d) ;\
-	cd $$GO_MOCKGEN_TMP_DIR ;\
-	go mod init tmp ;\
-	go get github.com/golang/mock/mockgen@v1.5.0 ;\
-	go install github.com/golang/mock/mockgen ;\
-	rm -rf $$GO_MOCKGEN_TMP_DIR ;\
-	}
+	hack/install-tools.sh github.com/golang/mock/mockgen@v1.5.0
 GO_MOCKGEN=$(shell go env GOPATH)/bin/mockgen
 else
 GO_MOCKGEN=$(shell which mockgen)

--- a/hack/install-tools.sh
+++ b/hack/install-tools.sh
@@ -1,0 +1,21 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+GO_MINOR_VERSION=$(go env GOVERSION | cut -c 2- | cut -d' ' -f1 | cut -d'.' -f2)
+function go::install() {
+  set -e
+  GO_INSTALL_TMP_DIR=$(mktemp -d)
+  cd "$GO_INSTALL_TMP_DIR"
+  go mod init tmp
+  if [ "$GO_MINOR_VERSION" -gt "16" ]; then
+    GO111MODULE=on go install "$@"
+  else
+    GO111MODULE=on go get "$@"
+  fi
+  rm -rf "$GO_INSTALL_TMP_DIR"
+}
+
+go::install "$@"

